x/mongo/driver/topology/sdam_spec_test.go

```diff
@@ -94,54 +120,275 @@ type testCase struct {
 
 const testsDir string = "../../../../data/server-discovery-and-monitoring/"
 
-func (r *response) UnmarshalJSON(buf []byte) error {
-	tmp := []interface{}{&r.Host, &r.IsMaster}
-	if err := json.Unmarshal(buf, &tmp); err != nil {
-		return err
+func (r *response) UnmarshalBSON(buf []byte) error {
+	doc := bson.Raw(buf)
+	if err := doc.Index(0).Value().Unmarshal(&r.Host); err != nil {
+		return fmt.Errorf("error unmarshalling Host: %v", err)
 	}
 
-	if len(tmp) != 2 {
-		return fmt.Errorf("'response' JSON array must have exactly two elements")
+	if err := doc.Index(1).Value().Unmarshal(&r.IsMaster); err != nil {
+		return fmt.Errorf("error unmarshalling IsMaster: %v", err)
 	}
 
 	return nil
 }
 
-func setUpFSM(t *testing.T, uri string) *fsm {
-	fsm := newFSM()
+func setUpTopology(t *testing.T, uri string) *Topology {
 	cs, err := connstring.ParseAndValidate(uri)
 	assert.Nil(t, err, "Parse error: %v", err)
 
-	fsm.SetName = cs.ReplicaSet
-	if fsm.SetName != "" {
-		fsm.Kind = description.ReplicaSetNoPrimary
-	}
-	if cs.DirectConnectionSet && cs.DirectConnection {
-		fsm.Kind = description.Single
-	}
+	topo, err := New(WithConnString(func(connstring.ConnString) connstring.ConnString {
+		return cs
+	}))
+	assert.Nil(t, err, "topology.New error: %v", err)
+
+	// add servers to topology without starting heartbeat goroutines
+	topo.serversLock.Lock()
+	for _, a := range topo.cfg.seedList {
+		addr := address.Address(a).Canonicalize()
+		topo.fsm.Servers = append(topo.fsm.Servers, description.Server{Addr: addr})
+
+		svr, err := NewServer(addr, topo.cfg.serverOpts...)
+		assert.Nil(t, err, "NewServer error: %v", err)
+		atomic.StoreInt32(&svr.connectionstate, connected)
+		svr.desc.Store(description.NewDefaultServer(svr.address))
+		svr.updateTopologyCallback.Store(topo.updateCallback)
 
-	for _, host := range cs.Hosts {
-		fsm.Servers = append(fsm.Servers, description.Server{Addr: address.Address(host).Canonicalize()})
+		topo.servers[addr] = svr
 	}
+	topo.desc.Store(description.Topology{Servers: topo.fsm.Servers})
+	topo.serversLock.Unlock()
 
-	return fsm
+	atomic.StoreInt32(&topo.connectionstate, connected)
+
+	return topo
 }
 
-func applyResponses(f *fsm, responses []response) error {
+func applyResponses(t *testing.T, topo *Topology, responses []response, sub *driver.Subscription) {
+	select {
+	case <-sub.Updates:
+	default:
+	}
 	for _, response := range responses {
 		doc, err := bson.Marshal(response.IsMaster)
-		if err != nil {
-			return err
+		assert.Nil(t, err, "Marshal error: %v", err)
+
+		addr := address.Address(response.Host)
+		desc := description.NewServer(addr, bsoncore.Document(doc))
+		server, ok := topo.servers[addr]
+		if ok {
+			server.updateDescription(desc)
+		} else {
+			topo.apply(context.Background(), desc)
```

Comment:
Why are we only applying to the topology if the server doesn't exist in `topo.servers`?

Reply:
update description calls topo.apply, but it also stores the new description on the server

Reply:
I'm a little confused here. In the case that the server is not in `topo.servers`, we call `topo.apply`, but the first thing it does is check if the description's address is in its `servers` map and returns without applying if it is not. What does calling `topo.apply` do here?

Reply:
there is a test that specifically checks for that behavior that still needs to go through topo.apply

Reply:
Can you add a comment explaining that? In our regular code, I don't think there's any path that updates the topology without going through `server.updateDescription`, so I find it weird that it's needed in this test.

Reply:
done